lib/pbench/test/unit/common/test_config.py

```diff
@@ -1,68 +1,252 @@
-        ), f"Unexpected object class for 'files', {config.files.__class__!r}"
-
-    def test_log_dir_provided(self):
-        config = PbenchConfig(_config_path_prefix / "logdir.cfg")
-        assert (
-            config.log_dir == "/srv/log/directory"
-        ), f"Unexpected log directory, {config.log_dir!r}"
-
-    def test_logger_type_provided(self):
-        config = PbenchConfig(_config_path_prefix / "hostport.cfg")
-        assert (
-            config.logger_type == "hostport"
-        ), f"Unexpected logger type, {config.logger_type!r}"
-        assert (
-            config.logger_host == "logger.example.com"
-        ), f"Unexpected logger host value, {config.logger_host!r}"
-        assert (
-            config.logger_port == "42"
-        ), f"Unexpected logger port value, {config.logger_port!r}"
-
-    def test_logger_type_hostport_missing(self):
-        with pytest.raises(BadConfig):
-            PbenchConfig(_config_path_prefix / "hostport-missing.cfg")
-        with pytest.raises(BadConfig):
-            PbenchConfig(_config_path_prefix / "hostport-missing-port.cfg")
+    def mock_file_list(self, cfg_name: str) -> List[str]:
+        """file_list mock module"""
+        return [cfg_name]
+
+    def test_default_logger(self, monkeypatch):
+        monkeypatch.setattr(configtools, "file_list", self.mock_file_list)
+        monkeypatch.setattr(ConfigParser, "__init__", MockedConfigParser.__init__)
+        monkeypatch.setattr(ConfigParser, "get", MockedConfigParser.get)
+        config = PbenchConfig("pbench.cfg")
+        assert config.logger_type == "devlog"
+
+    def test_logger(self, monkeypatch):
+        """verify logger_type value from config file"""
+
+        class mock_MockedConfigParser(MockedConfigParser):
+            def __init__(self):
+                self._sections = {"logging": {"logger_type": "file"}}
+
+        monkeypatch.setattr(configtools, "file_list", self.mock_file_list)
+        monkeypatch.setattr(ConfigParser, "__init__", mock_MockedConfigParser.__init__)
+        monkeypatch.setattr(ConfigParser, "get", MockedConfigParser.get)
+        config = PbenchConfig("pbench.cfg")
+        assert config.logger_type == "file"
+
+    def test_logger_type_hostport(self, monkeypatch):
+        """verify logger_type value as `hostport`, it expects `host` and `port`"""
+
+        class mock_MockedConfigParser(MockedConfigParser):
+            def __init__(self):
+                self._sections = {"logging": {"logger_type": "hostport"}}
+
+        monkeypatch.setattr(configtools, "file_list", self.mock_file_list)
+        monkeypatch.setattr(ConfigParser, "__init__", mock_MockedConfigParser.__init__)
+        monkeypatch.setattr(ConfigParser, "get", MockedConfigParser.get)
+        with pytest.raises(BadConfig) as exc:
+            config = PbenchConfig("pbench.cfg")
+            assert config.logger_type == "hostport"
+            assert exc == "No option 'logger_host' in section: 'logging'"
```

Comment:
There's a problem here:  the call to `PbenchConfig()` is going to raise the `BadConfig` exception, which means that execution is going to skip over the assertions at lines 72 & 73.

Line 73 needs to be moved out of the `with` block (and, `exc` is an exception-info object, so I think you have to extract the message from it before you compare it).

Line 72 should be removed: since the initialization of the configuration results in an exception, there's no reason to think that its contents have any particular values, so there's no point to checking them.

An analogous comment applies to each of the other `with` blocks.

Reply:
Good catch ... however even though I somehow missed the missed asserts here, I *did* notice the use of the `ExcInfo` object, and that's actually OK because it stringifies to the embedded exception text exactly as expected here.

(Although, yes, something like `str(exc.value)` would make that more explicit.)

Reply:
right

Reply:
Unfortunately, @dbutenhof, I'm not sure that `str(exc)` produces the same result as `str(exc.value)` (or, at least, they don't produce the result that I as a human expect...); e.g., note the expected message in this snippet from this PR:
```
        with pytest.raises(ValueError) as exc:
            PbenchConfig("pbench.cfg")
        expected_error_msg = "ValueError('logger_port', 'logging')"
        assert expected_error_msg in str(exc)
```
That message looks like the `repr(exc.value)` value.  `str(exc.value)` would be just the values (in this rather weird case, it's a tuple).  Me, I might have expected the message to look something like `ValueError: ('logger_port', 'logging')`...but I'm thinking that `ValueError` (and our (mis)use of it) might be a bad example to be quibbling over.  😬